src/literacy.rs

```diff
@@ -0,0 +1,123 @@
+#[cfg(all(feature = "std", feature = "use-core2"))]
+compile_error!("feature \"std\" and \"use-core2\" cannot be enabled together.");
+
+
+pub trait Read{
+    type Error;
+    fn read(&mut self, buf: &mut [u8]) -> ::core::result::Result<usize, Self::Error>;
+}
+
+pub trait Write {
+    type Error;
+    fn write(&mut self, buf: &[u8]) -> ::core::result::Result<usize, Self::Error>;
+    fn flush(&mut self) -> ::core::result::Result<(), Self::Error>;
+    fn write_all(&mut self, mut buf: &[u8]) -> ::core::result::Result<(), Self::Error> {
+        while !buf.is_empty() {
+            match self.write(buf) {
+                /*Ok(0) => {
+                    return Err(Error::new(ErrorKind::WriteZero, "failed to write whole buffer"));
+                }*/
+                Ok(n) => buf = &buf[n..],
+                //Err(ref e) if e.kind() == ErrorKind::Interrupted => {}
+                Err(e) => return Err(e),
+            }
+        }
+        Ok(())
+    }
+}
+
+#[cfg(feature = "std")]
+mod std_impl {
+    use super::{Read, Write};
+
+    impl<R: ::std::io::Read> Read for R {
```

Comment:
Is this what we want? Don't we want to `impl<R: OurReadTrait> std::io::Read for R` instead? That way a user can say `Hasher as std::io::Read`?

Reply:
Right, I'd managed to confuse myself between Read and Readable-style traits. In general I agree this is correct, but I do feel like its somewhat opposite of what we want for bitcoin_hashes. Maybe this is correct for Read but we should swap it for Write?

Reply:
> I feel the same for hashes, but thinking about rust-bitcoin and the encoding traits that would be annoying again.

Yea, I guess I'm ok with it as-is.

> Thinking about it, what keeps us from just implementing both std::io::Write and core2::Write for hash engines directly (depending on features), just calling the update fn.

For bitcoin_hashes, I don't disagree, but we'll want this stuff for rust-bitcoin (and rust-lightning), so I think its best to do it here and then use the same traits downstream.